src/include/index/ivf_pq_index.h

```diff
@@ -1244,16 +1281,141 @@ class ivf_pq_index {
         std::move(*generate_query_to_pq_centroid_distance_tables<
                   Q,
                   ColMajorMatrix<float>>(query_vectors));
-    return detail::ivf::query_finite_ram(
-        *partitioned_pq_vectors,
-        query_to_pq_centroid_distance_tables,
-        active_queries,
-        k_nn,
-        upper_bound,
-        num_threads_,
-        make_pq_distance_query_to_pq_centroid_distance_tables<
-            std::span<float>,
-            decltype(pq_storage_type{}[0])>());
+    size_t k_initial = static_cast<size_t>(k_nn * k_factor);
+    auto&& [initial_distances, initial_ids, initial_indices] =
+        detail::ivf::query_finite_ram(
+            *partitioned_pq_vectors,
+            query_to_pq_centroid_distance_tables,
+            active_queries,
+            k_initial,
+            upper_bound,
+            num_threads_,
+            make_pq_distance_query_to_pq_centroid_distance_tables<
+                std::span<float>,
+                decltype(pq_storage_type{}[0])>());
+    return re_rank(
+        std::move(initial_distances),
+        std::move(initial_ids),
+        std::move(initial_indices),
+        query_vectors,
+        k_initial,
+        k_nn);
+  }
+
+  auto re_rank(
+      ColMajorMatrix<float>&& initial_distances,
+      ColMajorMatrix<id_type>&& initial_ids,
+      ColMajorMatrix<size_t>&& initial_indices,
+      const auto& query_vectors,
+      size_t k_initial,
+      size_t k_nn) {
+    if (k_initial == k_nn) {
+      return std::make_tuple(
+          std::move(initial_distances), std::move(initial_ids));
+    }
+
+    if (::num_vectors(feature_vectors_) == 0 && !group_) {
```

Comment:
Should we also load all `feature_vectors_` for the infinite memory case during the constructor?

Reply:
Yep so that will happen (we hit that code in the second call to `rerank_query()` in `rerank()`. 

But this was just to cover the case where you create an empty index and then try to query with re-ranking. But I realized we actually don't need to throw here, so updated by moving the check for `group_` below and only reading from the URI if we have the group. Also added a test for this case.

Reply:
Sorry but I still don't understand where we load the full `feature_vectors_` array. Shouldn't this be inside `read_index_infinite`?

Reply:
Oh, oops, I thought you meant loading `feature_vectors_` during `add()`. You are correct, we do not load the full `feature_vectors_` anywhere except during `add()`.

In the constructor we don't know if we're opening for infinite or finite queries, that is per-query. We could perhaps load the full `feature_vectors_` during the call to `query_infinite()` (which calls `read_index_infinite()`), but then we will perhaps be loading way more vectors than we actually need to, and than we would if we just loaded them in `rerank()`. The benefit though would be we don't need to do any load during the second call to `query_infinite()`, whereas today we do.

Reply:
There are multiple options here, so maybe add some query/constructor parameters to make sure that we can leverage all of them and understand what happens in each case.

Reply:
Talked offline - will do this along with moving `memory_budget` into the index constructor in a follow-up.

Reply:
Please add a TODO or shortcut for this

Reply:
Started on it here! #518